File: 20_07_July/0729_42890_후보키/Solution.java
import java.util.ArrayList;
import java.util.HashSet;

class Solution {

    private boolean isUniqueColumns(String[][] relation, int columns) {
        HashSet<Integer> set = new HashSet<>();

        for (String[] record : relation) {
            int tmp = 0;

            for (int i = 0; i < record.length; i++) {
                // 해당 칼럼이 선택된 경우, set에 포함
                if ((columns & (1 << i)) != 0) {
                    tmp += record[i].hashCode();
                    tmp <<= 1;
                }
            }

            if (set.contains(tmp)) {
                // 이미 존재하는 record가 있으므로 유일성 만족X
                return false;
            }

            set.add(tmp);
        }

        return true;
    }

    private boolean isMinimum(ArrayList<Integer> results, int columns) {

        for (int result : results) {
            // or 연산으로 i가 result에 포함되어 있는지 확인
            if ((columns | result) == columns) {
                return false;
            }
        }

        return true;
    }

    public int solution(String[][] relation) {
        int columns = relation[0].length;
        int limit = 1 << columns;
        ArrayList<Integer> results = new ArrayList<>();

        for (int i = 1; i <= limit; i++) {
            // 최소성 만족 검사
            if (isMinimum(results, i) == false) {
                continue;
            }

            // 유일성 만족 검사
            if (isUniqueColumns(relation, i)) {
                results.add(i);
            }
        }

        return results.size();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.solution(new String[][] {{"100", "ryan", "music", "2"}, {"200", "apeach", "math", "2"}, {"300", "tube", "computer", "3"}, {"400", "con", "computer", "4"}, {"500", "muzi", "music", "3"}, {"600", "apeach", "music", "2"}}));
    }
}

File: 20_07_July/0729_42890_후보키/Solution.py

def solution(relation):

    # 각 컬럼이 희소성을 만족하는지 검사
    def is_unique(columns):
        s = set()

        for row in relation:
            tup = tuple(row[c] for c in columns)

            if tup in s:
                return False
            else:
                s.add(tup)

        return True

    # 현재 column이 최소성을 만족하는지 검사 ()
    def is_minimum(columns):
        
        for a in answer:
            if columns & a >= a:
                return False
        
        return True

    comb = 1
    l = len(relation[0])
    answer = []

    while comb < 2 ** l:
        # 10진(comb)값을 2진 배열로 변환한다.
        columns = [i for i in range(l) if comb & (1 << i) != 0]

        # 최소성, 유일성 모두 만족하는 경우
        if is_minimum(comb) and is_unique(columns):
            answer.append(comb)

        comb += 1

    return len(answer)


# exptected 2
print(solution([["100","ryan","music","2"],["200","apeach","math","2"],["300","tube","computer","3"],["400","con","computer","4"],["500","muzi","music","3"],["600","apeach","music","2"]]))
